runtime/plugin/matchparen.vim

```diff
@@ -1,225 +1,297 @@
-  " This fails when a timeout is not supported.
-  if mode() == 'i' || mode() == 'R'
-    let timeout = exists("b:matchparen_insert_timeout") ? b:matchparen_insert_timeout : g:matchparen_insert_timeout
+# RemoveHighlight {{{2
+
+if has('textprop')
+
+  def RemoveHighlight()
+    # `:silent!` to suppress E16 in case `line('w$')` is 0
+    silent! prop_remove({type: 'matchparen', all: true}, line('w0'), line('w$'))
+  enddef
+
+else
+
+  def RemoveHighlight()
+    if get(w:, 'matchparen')
+      silent! matchdelete(w:matchparen)
+      w:matchparen = 0
+    endif
+  enddef
+
+endif
+
+# Highlight {{{2
+if has('textprop')
+
+  # use text properties if available
+  def Highlight()
+    var props: dict<any> = {length: 1, type: 'matchparen'}
+    prop_add(c_lnum, c_col - before, props)
+    prop_add(m_lnum, m_col, props)
+  enddef
+
+else
+
+  # fall back on matchaddpos() otherwise
+  def Highlight()
+    w:matchparen = matchaddpos('MatchParen', [
+      [c_lnum, c_col - before],
+      [m_lnum, m_col]
+    ])
+  enddef
+
+endif
+
+def Toggle(enable = !exists('#matchparen')) #{{{2
+  if enable
+    Autocmds(true)
+    ParseMatchpairs()
+    UpdateHighlight()
   else
-    let timeout = exists("b:matchparen_timeout") ? b:matchparen_timeout : g:matchparen_timeout
+    Autocmds(false)
+    RemoveHighlight()
   endif
-  try
-    let [m_lnum, m_col] = searchpairpos(c, '', c2, s_flags, s_skip, stopline, timeout)
-  catch /E118/
-    " Can't use the timeout, restrict the stopline a bit more to avoid taking
-    " a long time on closed folds and long lines.
-    " The "viewable" variables give a range in which we can scroll while
-    " keeping the cursor at the same position.
-    " adjustedScrolloff accounts for very large numbers of scrolloff.
-    let adjustedScrolloff = min([&scrolloff, (line('w$') - line('w0')) / 2])
-    let bottom_viewable = min([line('$'), c_lnum + &lines - adjustedScrolloff - 2])
-    let top_viewable = max([1, c_lnum-&lines+adjustedScrolloff + 2])
-    " one of these stoplines will be adjusted below, but the current values are
-    " minimal boundaries within the current window
-    if i % 2 == 0
-      if has("byte_offset") && has("syntax_items") && &smc > 0
-	let stopbyte = min([line2byte("$"), line2byte(".") + col(".") + &smc * 2])
-	let stopline = min([bottom_viewable, byte2line(stopbyte)])
-      else
-	let stopline = min([bottom_viewable, c_lnum + 100])
-      endif
-      let stoplinebottom = stopline
+enddef
+
+def InStringOrComment(): bool #{{{2
+  # Expression which detects whether the current cursor position is in certain
+  # syntax types (string, comment, etc.); evaluated inside lambda passed as
+  # skip argument to searchpairpos().
+  return synstack('.', col('.'))
+    ->mapnew((_, v: number): string => synIDattr(v, 'name'))
+    # We match "escape" and "symbol" for special items, such as
+    # lispEscapeSpecial or lispBarSymbol.
+    ->match('\cstring\|character\|singlequote\|escape\|symbol\|comment') >= 0
```

Comment:
Doesn't this iterate through the entire list every time, instead of stopping when the first match is found?

Reply:
`match()` returns the index of the first item where the pattern matches:

   > When {expr} is a |List| then this returns the index of the
   > first item where {pat} matches.  Each item is used as a

To get the index of the first item, it doesn't need to iterate over the entire list (unless the pattern never matches).

Here is a simple test:
```vim
vim9script
var size: number = 10
var freq: number = 100'000
var hugelist: list<string> = repeat(['stack'], 100) + ['needle'] + repeat(['stack'], size)
def Test()
    for i in range(freq)
        eval match(hugelist, 'needle')
    endfor
enddef
defcompile
var time = reltime()
Test()
(reltime(time)->reltimestr()->matchstr('.*\..\{,3}') .. ' seconds to run Test()')->setline(1)
```
`size` controls the size of the list.
`freq` controls how frequently the function is invoked.

On my machine, this snippet reports that the function takes about 0.6s, no matter the value of `size`.
That is, even if I increase `size` from 10 to 100000, the function still takes 0.6s.  If `match()` was iterating over the entire list, it would take more time as the list grows in size; but it does not.

Reply:
Sorry, I've highlighted the wrong line. I should have highlighted the `mapnew()` instead of `match()`.

Regarding the `mapnew()` on line 244:

It seems that `synIDattr()` is called on every item returned by `synstack('.', col('.'))`. Shouldn't `synIDattr()` be called as we go through the list instead of calling it on every single item in the list? In other words, shouldn't it be something like this:

```
function InStringOrComment()
  for synID in synstack(line('.'), col('.'))
    if synIDattr(synID, 'name') =~? '\cstring\|character\|singlequote\|escape\|symbol\|comment'
      return 1
    endif
  endfor
  return 0
endfunction
```

This way, the number of calls to `synIDattr()` is minimized.

Reply:
I didn't notice a speed increase during a simple quick test, but I guess you're right, so I've pushed a commit which refactors the `mapnew()` into a `:for` loop.